server.go

```diff
@@ -367,7 +371,6 @@ func (srv *Server) Shutdown() error {
 		return &Error{err: "server not started"}
 	}
 	srv.started = false
-	srv.lock.Unlock()
```

Comment:
Same here, just `defer srv.lock.Unlock()` after we acquired the lock. If you do that I think the rest of the function can actually stay the same.